packages/txservice/src/provider.ts

```diff
@@ -0,0 +1,177 @@
+import { jsonifyError } from "@connext/nxtp-utils";
+import axios from "axios";
+import { BigNumber, Signer, Wallet, providers } from "ethers";
+import PriorityQueue from "p-queue";
+import { BaseLogger } from "pino";
+
+import { TransactionServiceConfig } from "./config";
+import { ChainError } from "./error";
+import { FullTransaction, MinimalTransaction, ProviderConfig } from "./types";
+
+const { JsonRpcProvider, FallbackProvider } = providers;
+
+/// Could use a more encompassing name, e.g. ChainRpcDispatch, etc
+export class ChainRpcProvider {
+  private provider: providers.JsonRpcProvider | providers.FallbackProvider;
+  private config: { gasInitialBumpPercent: number; gasPriceMinimum: BigNumber; rpcProviderMaxRetries: number; };
+  private signer: Signer;
+  private queue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  public confirmationsRequired: number;
+  public chainId: number;
+
+  constructor(
+    private readonly log: BaseLogger,
+    chainId: number,
+    signer: string | Signer,
+    providerConfigs: ProviderConfig[],
+    config: TransactionServiceConfig,
+  ) {
+    const { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries } = config;
+    this.config = { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries };
+    this.confirmationsRequired = config.chainConfirmations.get(chainId) ?? config.defaultConfirmationsRequired;
+    this.chainId = chainId;
+    
+    // Register a provider for each url.
+    // Make sure all providers are ready()
+
+    if (providerConfigs.length === 1) {
+      this.provider = new JsonRpcProvider(providerConfigs[0]);
+    } else {
+      const fallbackProviderConfigs = providerConfigs.map((config) => ({
+        provider: new JsonRpcProvider(
+          {
+            url: config.url,
+            user: config.user,
+            password: config.password,
+          }
+        ),
+        priority: config.priority ?? 1,
+        weight: config.weight ?? 1,
+        stallTimeout: config.stallTimeout,
+      }));
+      // TODO: Quorum is set to 1 here, but we may want to reconfigure later. Normally it is half the sum of the weights,
```

Comment:
That poses a very strange dilemma, but should be simple enough to solve. Have quorum be set by some optional argument passed into constructor (possibly a bool flag to indicate whether to set the quorum as half the sum of weights or just leave it as 1), then we will have our read/write classes instantiate providers differently

Reply:
Actually, circling back, this may be difficult. FallbackProvider doesn't allow for a change in quorum. So we can't just update the quorum if we're doing a read, and change it back for write.

Meaning we will need sep. provider instances for different quorums, and a write/read class would need to have access to both instances... which is weird

Reply:
Hmm, interesting. Let's just stick with the first provider to return strategy for now. Hopefully the `ready`/`synced` stuff will catch non-synced providers.